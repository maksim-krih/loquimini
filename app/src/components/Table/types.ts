import { GridPager, GridSearch } from "../../services/types";

export interface IProps {
  onRow: (record: any) => any;
  columns: any[];
  data: any[];
  pager: GridPager;
  loading: boolean;
  setPager: (_: any) => void;
  getData: (_: any, sorter: any, search?: GridSearch) => void;
}