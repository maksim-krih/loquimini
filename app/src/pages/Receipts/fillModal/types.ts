export interface IProps {
  visible: boolean;
  receiptId: string;
  onCancel: () => void;
}