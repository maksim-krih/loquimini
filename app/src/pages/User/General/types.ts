export interface IProps {
  isCreate?: boolean;
}
