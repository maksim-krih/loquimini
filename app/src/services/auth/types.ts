import { User } from "../types";

export interface Login {
  email: string;
  password: string;
}

export interface LoginResponse {
  user: User;
}