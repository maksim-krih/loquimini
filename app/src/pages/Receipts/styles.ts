import {createUseStyles} from 'react-jss'

export const useStyles = createUseStyles({
  container: {
    
  },
  gridButton: {
    padding: 0,
    width: 34,
    borderRadius: 5
  }
})