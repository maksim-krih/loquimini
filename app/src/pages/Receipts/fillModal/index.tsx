import React, { FC, useEffect, useState } from "react";
import { IProps } from "./types";
import { useStyles } from "./styles";
import { InputNumber, Modal } from "antd";
import Api from "../../../services";

const FillModal: FC<IProps> = (props: IProps) => {
  const classes = useStyles();
  const { visible, onCancel, receiptId, reloadGrid } = props;
  const [value, setValue] = useState<number | undefined>(undefined);

  const handleOk = () => {
    Api.Receipt.fillReceipt({
      newIndicator: value!,
      receiptId
    })
    .then((response: any) => {
      reloadGrid();
      onCancel();
    })
    .catch(e => {
      
    });
  }

  const onChange = (value: number) => {
    setValue(value);
  }

  return (
    <Modal 
      title="Fill Receipt" 
      visible={visible} 
      onOk={handleOk} 
      onCancel={onCancel}
    >
      <InputNumber value={value} onChange={onChange} style={{ borderRadius: 5 }}/>
    </Modal>
  );
}

export default FillModal;