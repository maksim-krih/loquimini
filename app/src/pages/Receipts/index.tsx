import React, { FC, useEffect, useState } from "react";
import { IProps } from "./types";
import { useStyles } from "./styles";
import { GridPager, GridSorter, Receipt } from "../../services/types";
import Api, { AuthService } from "../../services";
import { Button, Table, TablePaginationConfig, Typography } from "antd";
import { DefaultGridRequest, DefaultPager } from "../../consts";
import { 
  HouseType, 
  HouseTypeLabel, 
  ReceiptStatus, 
  ReceiptStatusLabel,
  ReceiptType, 
  ReceiptTypeLabel 
} from "../../enums";
import FillModal from "./fillModal";
import PayModal from "./payModal";
import { CheckCircleOutlined, CloseCircleOutlined, DollarCircleOutlined, EditOutlined, ExclamationCircleOutlined } from "@ant-design/icons";

const { Link } = Typography;

const Receipts: FC<IProps> = (props: IProps) => {
  const classes = useStyles();
  const [data, setData] = useState([]);
  const [fillModalVisible, setFillModalVisible] = useState(false);
  const [payModalVisible, setPayModalVisible] = useState(false);
  const [receiptId, setReceiptId] = useState("");
  const [pager, setPager] = useState(DefaultPager);
  const [loading, setLoading] = useState(false);

  const getStatus = (status: ReceiptStatus) => {
    switch (status) {
      case ReceiptStatus.Created:
        return (
          <div>
            <span><CloseCircleOutlined /></span>
            <span>{ReceiptStatusLabel(status)}</span>
          </div>
        );
      case ReceiptStatus.Filled:
        return (
          <div>
            <span><ExclamationCircleOutlined /></span>
            <span>{ReceiptStatusLabel(status)}</span>
          </div>
        );
      case ReceiptStatus.Paid:
        return (
          <div>
            <span><CheckCircleOutlined /></span>
            <span>{ReceiptStatusLabel(status)}</span>
          </div>
        );
    }
  }
  
  const columns = [
    {
      title: 'Date',
      dataIndex: 'date',
      sorter: true,
      render: (date: Date, record: Receipt) => (
        new Date(date).toLocaleDateString(undefined, {
          month: "long",
          year: "numeric",
        })
      )
    },
    {
      title: 'Address',
      dataIndex: 'address',
      sorter: true,
    },
    {
      title: 'House Type',
      dataIndex: 'houseType',
      sorter: true,
      render: (houseType: HouseType, record: Receipt) => (
        HouseTypeLabel(houseType)
      )
    },
    {
      title: 'Receipt Type',
      dataIndex: 'type',
      sorter: true,
      render: (type: ReceiptType, record: Receipt) => (
        ReceiptTypeLabel(type)
      )
    },
    {
      title: 'Status',
      dataIndex: 'status',
      sorter: true,
      render: (status: ReceiptStatus, record: Receipt) => (
        ReceiptStatusLabel(status)
      )
    },
    {
      title: 'Old Indicator',
      dataIndex: 'oldIndicator',
      sorter: true,
      render: (oldIndicator: number, record: Receipt) => {
        return record.type === ReceiptType.ColdWater ||
          record.type === ReceiptType.Electricity ||
          record.type === ReceiptType.HotWater ||
          record.type === ReceiptType.Gas
        ? oldIndicator : "—";
      }
    },
    {
      title: 'New Indicator',
      dataIndex: 'newIndicator',
      sorter: true,
      render: (newIndicator: number, record: Receipt) => {
        return record.type === ReceiptType.ColdWater ||
          record.type === ReceiptType.Electricity ||
          record.type === ReceiptType.HotWater ||
          record.type === ReceiptType.Gas
        ? newIndicator : "—";
      }
    },
    {
      title: 'Rate',
      dataIndex: 'rate',
      sorter: true,
      render: (rate: number, record: Receipt) => {
        return `$${rate}`;
      }
    },
    {
      title: 'Debt',
      dataIndex: 'debt',
      sorter: true,
      render: (debt: number, record: Receipt) => {
        return `$${debt}`;
      }
    },
    {
      title: 'Total',
      dataIndex: 'total',
      sorter: true,
      render: (total: number, record: Receipt) => {
        return `$${total}`;
      }
    },
    {
      title: 'Paid',
      dataIndex: 'paid',
      sorter: true,
      render: (paid: number, record: Receipt) => {
        return `$${paid}`;
      }
    },
    {
      title: '',
      dataIndex: '',
      render: (_: any, record: Receipt) => {
        switch (record.status) {
          case ReceiptStatus.Created:
            return (
              <Button 
                onClick={(e) => {
                  setReceiptId(record.id);
                  setFillModalVisible(true);
                }}
                className={classes.gridButton}
              >
                <EditOutlined />
              </Button>
            );
          case ReceiptStatus.Filled:
            return (
              <Button 
                onClick={(e) => {
                  setReceiptId(record.id);
                  setPayModalVisible(true);
                }}
                className={classes.gridButton}
              >
                <DollarCircleOutlined />
              </Button>
            );
        }
      },
    }
  ];

  useEffect(() => {
    getReceiptsData(pager);
  }, []);

  const getReceiptsData = (pager: GridPager, sorter?: GridSorter) => {
    setLoading(true);

    const request = DefaultGridRequest(pager, sorter);

    Api.Receipt.getByUserIdGrid(AuthService.User.id, request)
    .then((response: any) => {
      setData(response.data);
      setPager({...pager, total: response.total});
    })
    .catch(e => {
      
    })
    .finally(() => {
      setLoading(false);
    });
  };

  const handleTableChange = (pagination: TablePaginationConfig, filters: any, sorter: any) => {
    setPager({...pager, current: pagination.current!});
    getReceiptsData({...pager, current: pagination.current!}, sorter);
  };

  const fillModalCancel = () => {
    setFillModalVisible(false);
  };

  const payModalCancel = () => {
    setPayModalVisible(false);
  };

  return (
    <div className={classes.container}>
      {fillModalVisible && (
        <FillModal 
          visible={fillModalVisible} 
          onCancel={fillModalCancel}
          receiptId={receiptId}
          reloadGrid={() => getReceiptsData(pager)}
        />
      )}
      {payModalVisible && (
        <PayModal 
          visible={payModalVisible} 
          onCancel={payModalCancel}
          receiptId={receiptId}
          reloadGrid={() => getReceiptsData(pager)}
        />
      )}
      <Table
        columns={columns}
        rowKey={(record: Receipt) => record.id}
        dataSource={data}
        pagination={pager}
        loading={loading}
        onChange={handleTableChange}
      />
    </div>
  );
}

export default Receipts;