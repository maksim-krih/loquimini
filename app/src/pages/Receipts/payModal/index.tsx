import React, { FC, useEffect, useState } from "react";
import { IProps } from "./types";
import { useStyles } from "./styles";
import { InputNumber, Modal } from "antd";
import Api from "../../../services";

const PayModal: FC<IProps> = (props: IProps) => {
  const classes = useStyles();
  const { visible, onCancel, receiptId, reloadGrid } = props;
  const [value, setValue] = useState<number | undefined>(undefined);

  const handleOk = () => {
    Api.Receipt.payReceipt({
      value: value!,
      receiptId
    })
    .then((response: any) => {
      reloadGrid();
      onCancel();
    })
    .catch(e => {
      
    });
  }

  const onChange = (value: number) => {
    setValue(value);
  }

  return (
    <Modal 
      title="Pay Receipt" 
      visible={visible} 
      onOk={handleOk} 
      onCancel={onCancel}
    >
      <InputNumber value={value} onChange={onChange}/>
    </Modal>
  );
}

export default PayModal;