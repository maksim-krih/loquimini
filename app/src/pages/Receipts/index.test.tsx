import Receipts from ".";
import FillModal from "./fillModal";
import PayModal from "./payModal";
import { shallow } from "enzyme";
import { IProps as IFillModalProps } from "./fillModal/types";
import { IProps as IPayModalProps } from "./payModal/types";
import { Elements } from "@stripe/react-stripe-js";
import { loadStripe } from "@stripe/stripe-js";

const stripePromise = loadStripe('pk_test_WC7EYFzpuVCbs8ONcfSrjNE800zRXJe3lw');
const setUpReceipts = () => shallow(<Receipts />);
const setUpFillModal = (props: IFillModalProps) => shallow(<FillModal {...props} />);
const setUpPayModal = (props: IPayModalProps) => shallow(
  <PayModal  {...props} />
);

jest.mock('@stripe/react-stripe-js', () => ({
  ...jest.requireActual('@stripe/react-stripe-js'),
  useStripe: () => ({ }),
  useElements:  () => ({ })
}));


describe("Receipts", () => {
  describe("Index", () => {
    it("should render Receipts Component", () => {
      const component = setUpReceipts();
      expect(component).toMatchSnapshot();
    });
  });

  describe("FillModal", () => {
    it("should render FillModal Component", () => {
      const component = setUpFillModal({
        onCancel: () => {},
        receiptId: "",
        reloadGrid: () => {},
        visible: false
      });
      expect(component).toMatchSnapshot();
    });
  });

  describe("PayModal", () => {
    it("should render PayModal Component", () => {
      const component = setUpPayModal({
        onCancel: () => {},
        receiptId: "",
        reloadGrid: () => {},
        visible: false
      });
      expect(component).toMatchSnapshot();
    });
  });
});