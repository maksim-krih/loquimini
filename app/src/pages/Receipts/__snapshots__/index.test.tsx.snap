// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Receipts FillModal should render FillModal Component 1`] = `
<Modal
  confirmLoading={false}
  okType="primary"
  onCancel={[Function]}
  onOk={[Function]}
  title="Fill Receipt"
  visible={false}
  width={520}
>
  <ForwardRef
    min={1}
    onChange={[Function]}
    style={
      Object {
        "borderRadius": 5,
      }
    }
  />
</Modal>
`;

exports[`Receipts Index should render Receipts Component 1`] = `
<div
  className="container-0-2-1"
>
  <Table
    columns={
      Array [
        Object {
          "dataIndex": "date",
          "render": [Function],
          "sorter": true,
          "title": "Date",
        },
        Object {
          "dataIndex": "address",
          "sorter": true,
          "title": "Address",
        },
        Object {
          "dataIndex": "houseType",
          "render": [Function],
          "sorter": true,
          "title": "House Type",
        },
        Object {
          "dataIndex": "type",
          "render": [Function],
          "sorter": true,
          "title": "Receipt Type",
        },
        Object {
          "dataIndex": "status",
          "render": [Function],
          "sorter": true,
          "title": "Status",
          "width": 100,
        },
        Object {
          "dataIndex": "oldIndicator",
          "render": [Function],
          "sorter": true,
          "title": "Old Indicator",
        },
        Object {
          "dataIndex": "newIndicator",
          "render": [Function],
          "sorter": true,
          "title": "New Indicator",
        },
        Object {
          "dataIndex": "rate",
          "render": [Function],
          "sorter": true,
          "title": "Rate",
        },
        Object {
          "dataIndex": "debt",
          "render": [Function],
          "sorter": true,
          "title": "Debt",
        },
        Object {
          "dataIndex": "total",
          "render": [Function],
          "sorter": true,
          "title": "Total",
        },
        Object {
          "dataIndex": "paid",
          "render": [Function],
          "sorter": true,
          "title": "Paid",
        },
        Object {
          "dataIndex": "",
          "render": [Function],
          "title": "",
        },
      ]
    }
    dataSource={Array []}
    loading={false}
    onChange={[Function]}
    pagination={
      Object {
        "current": 1,
        "pageSize": 10,
        "total": 0,
      }
    }
    rowKey={[Function]}
  />
</div>
`;

exports[`Receipts PayModal should render PayModal Component 1`] = `
<ContextProvider
  value={
    Object {
      "elements": null,
      "stripe": null,
    }
  }
>
  <PayModal
    onCancel={[Function]}
    receiptId=""
    reloadGrid={[Function]}
    visible={false}
  />
</ContextProvider>
`;
